main.go

```diff
@@ -41,15 +40,16 @@ func main() {
 		log.Fatal("Error finding and binding port:", err)
 	}
 
+	// Goroutine to handle shutdown signals
 	go func() {
-		<-sigs
-		log.Info("Shutdown signal received")
-		cancel()
+		log.Info("Waiting for shutdown signal")
+		asyncServer.WaitForSignal(cancel, sigs)
```

Comment:
We can invoke asyncServer.InitiateShutdown() and avoid adding a new function WaitForSignal.
I think the reason the port wasn't getting closed was because cancel() forced all the go-routines to stop immediately and the program exited before the port was closed cleanly.

Reply:
Makes sense, fixed it and verified the functionality.